src/hdbscan.jl

```diff
@@ -0,0 +1,296 @@
+        end
+    end
+    return minimum_spanning_tree
+end
+
+function build_hierarchy(mst, min_size)
+    n = length(mst) + 1
+    cost = 0
+    uf = UnionFind(n)
+    Hierarchy = Array{HdbscanCluster}(undef, 2n-1)
+    if min_size == 1
+        for i in 1 : n
+            Hierarchy[i] = HdbscanCluster(false, [i])
+        end
+    else
+        for i in 1 : n
+            Hierarchy[i] = HdbscanCluster(true, Int[])
+        end
+    end
+    sort!(mst)
+    
+    for i in 1 : n-1
+        c, j, k = mst[i]
+        cost += c
+        λ = 1 / cost
+        #child clusters
+        c1 = group(uf, j)
+        c2 = group(uf, k)
+        #reference to the parent cluster
+        Hierarchy[c1].parent = Hierarchy[c2].parent = n+i
+        nc1, nc2 = isnoise(Hierarchy[c1]), isnoise(Hierarchy[c2])
+        if !(nc1 || nc2)
+            #compute stability
+            compute_stability(Hierarchy[c1], λ)
+            compute_stability(Hierarchy[c2], λ)
+            #unite cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            Hierarchy[n+i] = HdbscanCluster(false, points)
+        elseif !(nc1 && nc2)
+            if nc2 == true
+                (c1, c2) = (c2, c1)
+            end
+            #record the lambda value
+            append!(Hierarchy[c2].λp, fill(λ, length(Hierarchy[c1])))
+            #unite cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            Hierarchy[n+i] = HdbscanCluster(false, points)
+        else
+            #unite the noise cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            if length(points) < min_size
+                Hierarchy[n+i] = HdbscanCluster(true, Int[])
+            else
+                Hierarchy[n+i] = HdbscanCluster(false, points)
+            end
+        end
+    end
+    return Hierarchy
+end
+
+function extract_cluster!(hierarchy::Vector{HdbscanCluster})
+    for i in 1 : length(hierarchy)-1
+        if isnoise(hierarchy[i])
+            c = hierarchy[i]
+            push!(hierarchy[c.parent].children, i)
+            hierarchy[c.parent].children_stability += c.stability
+        else
+            c = hierarchy[i]
+            if c.stability > c.children_stability
+                push!(hierarchy[c.parent].children, i)
+                hierarchy[c.parent].children_stability += c.stability
+            else
+                append!(hierarchy[c.parent].children, c.children)
+                hierarchy[c.parent].children_stability += c.children_stability
+            end
+        end
+    end
+end
+
+# Below are utility functions for building hierarchical trees
+heappush!(h, v) = insert!(h, searchsortedfirst(h, v), v)
+
+mutable struct UnionFind{T <: Integer}
+    parent:: Vector{T}  # parent[root] is the negative of the size
+    label::Dict{Int, Int}
+    cnt::Int
```

Comment:
Can you please put some comments that describe what this function does and what are its fields?
I have not tried too hard to understand what it, but ideally that should be clear from the comments.
The only field that could be modified is `cnt`, but IIUC `length(uf.label)` gets you the same result, so you can just remove `cnt` and turn the type into `struct`.

```suggestion
struct UnionFind{T <: Integer}
    parent:: Vector{T}  # parent[root] is the negative of the size
    label::Dict{Int, Int}
```

Reply:
I found this isn't right. The `cnt` field is used to count the number of unite and determine the new group id.
So, I changed the field name into `next_id` for clarity.